### ISSUE TITLE
switch exhaustiveness waiver for the null-state input is lost when a case is matched conditionally in a tuple pattern

### ISSUE THREAD

**Issue by OnurGumus** [dotnet/roslyn#84100]
**Version Used**: Roslyn 5.8.0-1.26302.115 (f7b4c5716faaee8fb8a289aed29118cad955c45f), .NET SDK 11.0.100-preview.5.26302.115, `<LangVersion>preview</LangVersion>`

### Summary

Switch exhaustiveness over a `union` waives the union's default-state input ("contains no case", matched by the `null` pattern) — both in a direct switch and inside a tuple pattern when every case is covered unconditionally. But as soon as one case is matched **conditionally** (its arm constrains another tuple element), the waiver is lost, and CS8655 surfaces the default-state input as the counterexample: `the pattern '(null, _)' is not covered`.

This is (a) inconsistent — whether the union's zero is waived shouldn't depend on conditions applied to *other* tuple elements — and (b) a misleading diagnostic: the actionable gap in such code is usually a genuine value gap (here `(B, null)`), which the message hides behind the exotic default-state counterexample.

### Steps to Reproduce

```csharp
public record A;
public record B;
public union U(A, B);

public static class Repro
{
    // 1. Direct union switch: no warning (default-state input is waived)
    public static string Direct(U u) => u switch { A => "a", B => "b" };

    // 2. Tuple pattern, every case covered unconditionally: no warning
    public static string Unconditional(U u, string? s) => (u, s) switch
    {
        (A, _) => "a",
        (B, _) => "b",
    };

    // 3. Tuple pattern, one case covered conditionally: CS8655
    public static string Conditional(U u, string? s) => (u, s) switch
    {
        (A, _) => "a",
        (B, { } t) => t,
    };
}
```

### Actual Behavior

Only method 3 warns:

```
warning CS8655: The switch expression does not handle some null inputs (it is not
exhaustive). For example, the pattern '(null, _)' is not covered.
```

Adding a `(null, _)` arm to method 3 keeps CS8655 but moves the example to the real value gap:

```
warning CS8655: ... For example, the pattern '(B, null)' is not covered.
```

So the uncovered set is `{ (null, _), (B, null) }`, and the diagnostic chooses to lead with the default-state union.

(The default-state input is real at runtime: `default(U)` matches none of the case patterns and *does* match the `null` pattern — consistent with "patterns apply to the union's contained value". The question is only about the waiver's scope and the counterexample choice.)

### Expected Behavior

One of:

1. **Waive consistently**: if the default-state input is waived for `Direct` and `Unconditional`, it should also be waived in `Conditional` — the union dimension's cases are equally enumerated in all three; only the *other* tuple element's coverage differs. Method 3 would then report the genuine value gap as CS8509 with `(B, null)`.
2. Or, if the default-state input is intentionally counted whenever the analysis leaves the union-aware path: prefer a non-default-state counterexample when one exists (`(B, null)` is strictly more actionable than `(null, _)`), and consider documenting that the waiver only applies to unconditional case coverage.

Related: #83666 (CS8509 over single-case union, fixed).

**Comment by AlekseyTs** [comment]
It looks like the behavior is not Unions specific. It can be observed with `bool?`, for example.
```
public static class Repro
{
    public static string Conditional(bool? u, string? s)
    {
        if (u is null) return "";

        u.Value.ToString();

        // (8,23): warning CS8655: The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '(null, _)' is not covered.
        return (u, s) switch
        {
            (true, _) => "a",
            (false, { } t) => t,
            //(null, _) => ""
            //(false, null) => ""
        };
    }
}
```

**Comment by RikkiGibson** [comment]
This might be a case where the `{ }` pattern is considered a "pure null test", i.e. you are expressing doubt about whether this thing may be null, so its flow state is adjusted to maybe-null unconditionally at that point. We'll want to double check.

A similar effect occurs for the following scenario

```cs
void M(string s)
{
    if (s == null) { }

    s.ToString(); // warning
}
```

**Comment by AlekseyTs** [comment]
What is interesting is that adding a null check for `s` makes the warning go away:
```
public static class Repro
{
    public static string Conditional(bool? u, string? s)
    {
        if (u is null) return "";
        if (s is null) return "";

        u.Value.ToString();

        return (u, s) switch
        {
            (true, _) => "a",
            (false, { } t) => t,
        };
    }
}
```
